src/deluge/model/mod_controllable/mod_controllable_audio.cpp

```diff
@@ -1681,137 +1681,368 @@ bool ModControllableAudio::offerReceivedCCToLearnedParams(MIDIDevice* fromDevice
-						newKnobPos = 64;
-						if (value < 127) {
-							newKnobPos = (int32_t)value - 64;
-						}
-						knob->previousPositionSaved = false;
+					//add 64 to internal knobPos to compare to midi cc value received
+					//if internal pos + 64 is greater than 127 (e.g. 128), adjust it to 127
+					//because midi can only send a max midi value of 127
+					int32_t knobPosForMidiValueComparison = knobPos + kKnobPosOffset;
+					if (knobPosForMidiValueComparison > 127) {
+						knobPosForMidiValueComparison = 127;
 					}
-					else { //Midi Takeover Mode = Pickup or Value Scaling
-						/*
 
-						Step #1: Convert Midi Controller's CC Value to Deluge Knob Position Value
+					//is the cc being received for the same value as the current knob pos? If so, do nothing
+					if (value != knobPosForMidiValueComparison) {
+						newKnobPos = calculateKnobPosForMidiTakeover(modelStackWithParam, knobPos, value, knob);
+					}
+					else {
+						continue;
+					}
+				}
 
-						- Midi CC Values for non endless encoders typically go from 0 to 127
-						- Deluge Knob Position Value goes from -64 to 64
+				//Convert the New Knob Position to a Parameter Value
+				int32_t newValue =
+				    modelStackWithParam->paramCollection->knobPosToParamValue(newKnobPos, modelStackWithParam);
 
-						To convert Midi CC Value to Deluge Knob Position Value, subtract 64 from the Midi CC Value
+				//Set the new Parameter Value for the MIDI Learned Parameter
+				modelStackWithParam->autoParam->setValuePossiblyForRegion(newValue, modelStackWithParam, modPos,
+				                                                          modLength);
 
-						So a Midi CC Value of 0 is equal to a Deluge Knob Position Value of -64 (0 less 64).
+				//if you're in automation view and editing the same parameter that was just updated
+				//by a learned midi knob, then re-render the pads on the automation editor grid
+				if (getRootUI() == &automationInstrumentClipView) {
+					InstrumentClip* clip = (InstrumentClip*)currentSong->currentClip;
+					int32_t id = modelStackWithParam->paramId;
+					Param::Kind kind = modelStackWithParam->paramCollection->getParamKind();
+					if ((clip->lastSelectedParamID == id) && (clip->lastSelectedParamKind == kind)) {
+						uiNeedsRendering(&automationInstrumentClipView);
+					}
+				}
+			}
+		}
+	}
+	return messageUsed;
+}
 
-						Similarly a Midi CC Value of 127 is equal to a Deluge Knob Position Value of +63 (127 less 64)
+/// this function is called when midi follow is enabled
+/// it checks if the ccNumber received has been learned to any parameters in midi learning view
+/// if the cc has been learned, it sets the new value for that parameter
+/// this function works by first checking the active context to see if there is an active clip
+/// to determine if the cc intends to control a song level or clip level parameter
+void ModControllableAudio::offerReceivedCCToMidiFollow(ModelStack* modelStack, Clip* clip, int32_t ccNumber,
```

Comment:
This is no longer required right? Since now it's being called from the playback handler

Reply:
as discussed, the function is required and to avoid confusion I've now renamed the function to "receivedCCFromMidiFollow" to accurately describe what the function is doing